readme.md
# Node api-quick


Create a simple JSON based API server in as little as 6 lines with built in support for *SSL*, *Basic auth* & *rate limiting*. Capable of handling up to **2200 requests/sec** on a single core ([more info](docs/performance.md)).


[![npm](https://img.shields.io/npm/v/api-quick.svg?maxAge=2592000)](https://www.npmjs.com/package/api-quick)
[![GitHub issues](https://img.shields.io/github/issues/jake314159/NodeApiQuick.svg)](https://github.com/jake314159/NodeApiQuick/issues)
[![travis](https://travis-ci.org/jake314159/NodeApiQuick.svg?branch=travis-setup)](https://travis-ci.org/jake314159/NodeApiQuick)
[![travis](https://david-dm.org/jake314159/NodeApiQuick.svg)](https://david-dm.org/)



## Install me

Install using [npm](https://www.npmjs.com/package/api-quick) with the command

```bash
npm install api-quick --save
```

## Basic API server

Time for a quick 6 line example! The below code creates an api server that responds to the port 8080 and returns the current date-time with the url "http://127.0.0.1:8080/date".

```javascript
var api = require('api-quick').init(8080);
var endpoints = {};
endpoints.date = function() {
  return {date: new Date().toUTCString()};
};
api.addEndpoints(endpoints);
```

Doing a GET request on the above url will then return the data:

```
{"date":"Tue, 17 May 2016 17:11:07 GMT"}
```

## Multi-level endpoints


It is also possible to have a longer multi-level url path with nested Objects as shown below

```javascript
var api = require('api-quick').init(8080);
var endpoints = {
    date: {
        utc: function() {
            return {date: new Date().toUTCString()};
        },
        iso: function() {
            return {date: new Date().toISOString()};
        },
    }
};
api.addEndpoints(endpoints);
```

GET http://127.0.0.1:8080/date/utc

```
{"date":"Sat, 11 Jun 2016 15:04:55 GMT"}
```

GET http://127.0.0.1:8080/date/iso

```
{"date":"2016-06-11T15:04:55.418Z"}
```

## Parameters

Another way of allowing different date formats might be to be specified might be my using url parameters, for example:

```javascript
var api = require('api-quick').init(8080);
var endpoints = {};
endpoints.date = function(req) {
    if(req.body.format === 'utc') {
        return {date: new Date().toUTCString()};
    } else {
        return {date: new Date().toISOString()};
    }
};
api.addEndpoints(endpoints);
```

GET http://127.0.0.1:8080/date?format=utc

```
{"date":"Sat, 11 Jun 2016 15:04:55 GMT"}
```

GET http://127.0.0.1:8080/date?format=iso

```
{"date":"2016-06-11T15:04:55.418Z"}
```


## Callbacks

The handler can also return the result by taking a callback function as an argument, for example:

```javascript
var api = require('api-quick').init(8080);
var endpoints = {};
endpoints.date = function(req, callback) {
    var data = {date: new Date().toISOString()};
    callback(null, data);
};
api.addEndpoints(endpoints);
```

Note that the first argument to the callback is an optional Error object and the second is the data to return.

## URL Layout

The components of the url that are provided to your function is shown below

Example url:
```
http://127.0.0.1:8080/<:function_route:>/<:args:>?<:params:>
```

Maps to the function arguments:
```javascript
function(req, callback)
```

+ **req.method**: 'GET' or 'POST'
+ **req.args**: An array of strings for the elements of the url not used to find the handler function.
+ **req.body**: is either the url encoded parameters for GET requests or the posted data for POST requests in a JSON format.




So for the basic example above the url:

```
http://127.0.0.1:8080/date/now/utc?format=json
```

will call the handler function with the following arguments:

```javascript
{
  'date': function(req, cb) {
    console.log(req.method); // == 'GET'
    console.log(req.args); // == ['now', 'utc']
    console.log(req.body); // == {'format': 'json'}
  }
}
```

The hander function should return a json structure to return to the client.

## SSL

To use SSL simply give the paths to the key and certificate files in the initialisation data.

```javascript
var api = require('api-quick');
api.init(8080,{
    'ssl': {
        'key': './key.pem',
        'cert': './cert.pem'
    }
});
```

## Rate limits

Also included is some basic rate limit functionality, disabled by default. There are two ways of doing this...

Using the default values:

```javascript
api.init(8080, {'rateLimit': true});
```

or using your own custom values:

```javascript
api.init(8080, {
    'rateLimit': {
        'period': 60, //Seconds
        'limit': 60
    }
});
```


## Basic auth

ApiQuick supports basic auth which is where a username and password is encoded with base64 and put into the '*Authorization*' header. An example of sending the username and password 'test' with curl is shown below:
```
curl -H "Authorization: Basic dXNlcjpwYXNz" 127.0.0.1:8080/date
```

The authentication works by running an authentication function that returns (via a callback) either *true* or *false* indicating if the username and password given is valid. There are a few method of doing this.

### One

By providing username and password pairs in a json format. This will be applied as a global auth function.

```javascript
api.authByJson({'username': 'password'});
```

Multiple password keys can also be supplied for one user with an array.

```javascript
api.authByJson({'username': ['key1', 'key2']});
```

### Two

By using a global auth function which applies to all endpoints.

```javascript
api.auth(function(user, pass, callback) {
  callback(pass == 'test');
});
```

### Three

By doing an endpoint specific function (supplied in the extra paramiter)

```javascript
api.addEndpoints(endpoints, {
  'auth': function(user, pass, callback) {
    callback(pass == 'passw0rd');
  }
});
```

Endpoint specific auth functions are used if present and if not then the global function is used. If there are no auth functions then all requests are authorised.

## Cache Control

The cache control object can take most/all of the [valid flags](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control), converted to camel case.
Cache option can be passed into the ```init``` function to apply it to all endpoints or to the ```addEndpoints``` method to apply only to specific endpoints.
Endpoint specific cacheControl settings will take priority over global cache settings.
Note that in debug mode caching will be disabled.


Here are some common examples:


**Don't cache any endpoints**

```javascript
api.init(8080, {
  'cacheControl': false
});
```

**Cache for a year**

```javascript
api.init(8080, {
  'cacheControl': 31557600
});
```

which is the same as

```javascript
api.init(8080, {
  'cacheControl': {
    public: true,
    maxAge: 31557600
  }
});
```

**Cache a single endpoint for 1 day client side but not in the cdn**

```javascript
api.addEndpoints(endpoints, {
  'cacheControl': {
    private: true,
    maxAge: 86400
  }
});
```

## Initialisation options

| Field      | Description                                               | Default |
|:---------- |:--------------------------------------------------------- | ------- |
| SSL        | Specifies SSL encryption settings (see above)             | false   |
| rateLimit  | Specifies rate limit settings (see above)                 | false   |
| prettyJson | Pretty print the JSON response data                       | false   |
| consoleLog | Log events to standard out for debugging                  | 'info'  |
| compress   | Compress all connections with gzip                        | false   |
| maxDepth   | Maximum number of url arguments to allow                  | 1       |
| debug      | Include extra debug information in responses              | false   |
| fullRequest| Include all the request information avalible              | false   |
| headers    | An object containing additional headers to add to all requests | {} |
| cacheControl | Control cache behaviour              | {public: true, maxAge: 30} |

## Dependencies

+ secure-compare
+ events
+ url
+ compression

## License

Copyright 2015 Jacob Causon

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.


CHANGELOG.md
# Release history

## Unreleased

+ Able to add custom headers to requests (for all requests and for specific endpoints) with the ```headers``` init option
+ Added option to change cacheControl header with '''cacheControl''' init option. **Documentation TODO**
+ Added ```Access-Control-Allow-Origin: *``` header by default as it's required for common use cases

## v0.3.3 (2017-04-02)

+ Reduce dependencies
+ Fix leaking x-powered-by header
+ Various bug fixes
+ More tests

## v0.3.2 (2016-06-14)
+ Performance benchmarks
+ Documentation improvements

## v0.3.1 (2016-06-12)

+ Fixup error in README
+ Expose client IP address in request data
+ Option to pass all request information to the handler
+ Performance benchmark conducted and details included in ```docs/performance.md```

## 0.3.0 BETA (2016-06-06)

+ Option for handler to return asynchronously if it takes a callback (but will behave as before if not) **(BREAKING)**
+ Authentication functions run asynchronously via a callback **(BREAKING)**
+ Only allow 1 argument by default **(BREAKING)**
+ Handlers function given a single object with all the request details **(BREAKING)**
+ Support for express compatible middleware
+ Gzip compression support
+ Pretty json output
+ Removed dependency on express
+ Infinite depth endpoint urls
+ Improved log event system
+ Logs to standard out by default

## 0.2.0 BETA (2016-05-21)

+ Mark *addPackage* as **deprecated**, due to be removed v1.0.0
+ Implement addEndpoints which will replace addPackage
+ Major code refactor
+ Tests for the full program
+ Disable auth by setting ```auth: false```
+ Fix dependency versions
+ Update readme


## 0.1.2 BETA (2016-05-17)

**Security update (LOW severity)**

+ Use a constant time string comparison for API keys as part of the authByJson function. Fixes a potential timing leak relating to the use of authByJson that could theoretically reduce the number of attempts required to brute force an api key.
+ Adds an extra requirement for 'secure-compare'

## 0.1.1 BETA

+ Fix a bug in the standard auth code

## 0.1.0 BETA

+ Initial commit


lib/ApiQuick.js
'use strict';

var events = require('events').EventEmitter;
var url = require('url');
var bodyParser = require('body-parser');
var compression = require('compression');

var middleware = require('./middleware');
var ratelimit = require('./ratelimit');
var Auth = require('./auth');

var ApiQuick = function () {};

ApiQuick.prototype = {

    // Constants
    error: 'error',
    warn: 'warn',
    info: 'info',

    /**
     * Initializes the api server
     * @param port: An optional port for the server to listen to, default: 8080
     * @param extra: An optional dict containing extra settings for the api server
     */
    init: function(port, extra) {
        var self = this;
        if(!extra) extra = {};

        this.logger = new events();
        this.middleware = new middleware();
        this.rateLimit = new ratelimit();

        if(extra.compress) {
            this.use(compression());
        }

        this.use(function(req, res, next) {
            res.removeHeader('X-Powered-By');
            next();
        });

        // Handle the rate limits with middleware
        if(extra.rateLimit) {
            this.rateLimit.startRateLimit(extra.rateLimit);
            this.use(function (req, res, next) {
                if (!self.rateLimit.handleRateLimit(req.connection.remoteAddress)) {
                    // Rate limit reached
                    self._writeResponse({ok: false, code: 429, error: 'Rate limit reached'});
                } else {
                    next();
                }
            });
        }

        this.use(bodyParser.json());
        this.use(bodyParser.urlencoded({
            extended: true
        }));

        // Init vars
        this.routes = {};
        this.checkAuth = false;
        this.ssl = false;
        this.options = {};
        this.port = port || 8080;
        this.prettyJson = false;
        this.consoleLog = 4;  // Log everything
        this.maxDepth = 1;
        this.debug = false;
        this.fullRequest = false;
        this.headers = {};
        this.cacheControl = null;

        // Deal with extra settings
        if(extra.ssl && extra.ssl.key && extra.ssl.cert) {
            var fs = require('fs');
            this.ssl = true;
            this.options.key = fs.readFileSync(extra.ssl.key);
            this.options.cert = fs.readFileSync(extra.ssl.cert);
        }
        if(extra.prettyJson !== undefined) {
            this.prettyJson = extra.prettyJson;
        }
        if(extra.maxDepth !== undefined) {
            this.maxDepth = extra.maxDepth;
        }
        if(extra.debug) {
            this.debug = extra.debug;
        }
        if(extra.fullRequest !== undefined) {
            this.fullRequest = extra.fullRequest;
        }
        if(extra.headers !== undefined) {
            this.headers = extra.headers;
        }
        if(extra.cacheControl !== undefined) {
            this.cacheControl = extra.cacheControl;
        }
        if(extra.consoleLog !== undefined) {
            this.consoleLog = extra.consoleLog;
        }

        if(this.consoleLog) {
            this._startConsoleLogging(this.consoleLog);
        }

        // Start the server
        if(this.ssl) {
            require('https').createServer(this.options, this._process()).listen(this.port);
        } else {
            require('http').createServer(this._process()).listen(this.port);
        }

        this.logger.emit('info', 'Listening to port ' + this.port, {});

        return this;
    },

    _startConsoleLogging: function(level) {
        // Convert a string debug value to an int value for easy comparison
        switch(level) {
            case 'error':
                level = 8;
                break;
            case 'warn':
                level = 6;
                break;
            case 'info':
                level = 4;
                break;
            case true:
                level = 2;
                break;
            default: // Leave it as it is
        }

        if(level <= 8) {
            this.logger.on(this.error, function(msg, data) {
                console.log('ERROR    ', new Date().toISOString(), '    ', msg, data);
            });
        }
        if(level <= 6) {
            this.logger.on(this.warn, function(msg, data) {
                console.log('WARN     ', new Date().toISOString(), '    ', msg, data);
            });
        }
        if(level <= 4) {
            this.logger.on(this.info, function(msg, data) {
                console.log('INFO     ', new Date().toISOString(), '    ', msg, data);
            });
        }
    },

    /**
     * Add express compatible middleware to the api server that will run for every connection
     * @param f: A function to run for every connection before it is handled by the api server
     *                 Function is given the parameters req, res, next.
     *                req: Express request object
     *                res: Express response object
     *                next: Callback function
     */
    use: function(f) {
        this.middleware.add(f);
    },

    /**
     * Add a listener for api server events
     * @params type: Type of event to listen for ('error', 'warn', 'info')
     */
    on: function(type, f) {
        this.logger.on(type, f);
    },

    /**
     * Adds a single specified endpoint to the api server
     * @deprecated: To be removed for v1.0.0
     * @param name: The first component of the endpoint url
     * @param p: The rest of the package dict
     * @param extra: Extra settings information for the provided dict
     */
    addPackage: function (name, p, extra) {
        var endPoints = {};
        endPoints[name] = p;
        this.addEndpoints(endPoints, extra);
    },

    /**
     * Takes a potentially multi-layer dict of functions and adds them as endpoints to the api server
     * Replaces addPackage
     * @param route: A potentially multi-layer dict of functions
     * @param extra: An optional dict of options to apply to the provided endpoints
     */
    addEndpoints: function(route, extra) {
        var stack = [];
        var self = this;
        function _r(d) {
            for(var k in d) {
                stack.push(k);
                if(typeof d[k] == 'function') {
                    if(extra && extra.auth) {
                        d[k].auth = extra.auth;
                    }
                    if(extra && extra.headers) {
                        d[k].headers = extra.headers;
                    }
                    if(extra && typeof extra.cacheControl !== "undefined") {
                        d[k].cacheControl = extra.cacheControl;
                    }
                    self.routes['/' + stack.join('/')] = d[k];
                } else if(typeof d[k] == 'object'){
                    //It's a dict to recursively search deeper
                    _r(d[k]);
                }
                stack.pop();
            }
        }

        _r(route);
    },

    /**
     * Takes a request object and checks that it has a valid handler
     * Will also return the handler if one can be found
     * @param req: The request object
     * @returns: A dict with an ok & code parameter and optionally an error string or a handler function
     */
    _checkParamsSupplied: function(req) {
        var endpoint = req.u.pathname;
        if(endpoint[endpoint.length-1] == '/') {
            // remove the end slash
            endpoint = endpoint.substring(0, endpoint.length-1);
        }

        var returnDict = {
            ok: false,
            code: 404,
            error: "No endpoint " + endpoint
        };

        var args = [];
        var depth = 0;

        // Look down the endpoint url until a valid handler is found
        // This is expensive when there is no handler so keep maxDepth as low as possible
        while(!returnDict.ok && depth <= this.maxDepth) {
            if(this.routes[endpoint]) {
                returnDict.ok = true;
                returnDict.handler = this.routes[endpoint];
                returnDict.handler.auth = returnDict.handler && returnDict.handler.auth
                returnDict.args = args;
                returnDict.error = undefined;
            } else {
                depth++;
                if(depth > this.maxDepth) break; // No point doing any more we have reached the limit
                var endIndex = endpoint.lastIndexOf('/');
                if(endIndex !== 0) {
                    args.unshift(endpoint.substring(endIndex+1, endpoint.length))
                    endpoint = endpoint.substring(0, endIndex);
                } else {
                    break;
                }
            }
        }
        return returnDict
    },

    /**
     * Returns the headers that should be added when sending the provided data
     * @param data: The data that the header will be sent with
     */
    _getHeaders: function(data, extra) {
        var header = {
            'content-type' : 'application/json',
            'X-Content-Type-Options': 'nosniff',
            'X-Frame-Options': 'DENY',
            'Access-Control-Allow-Origin': '*'
        };
        if(data && data.code == 401) {
            // Message was an auth fail so we want to give a hint on how to auth
            header['WWW-Authenticate'] =  'Basic user:pass';
        }

        // Handle cacheing
        var cacheObj = this.cacheControl
        if(extra && extra.handler && typeof extra.handler.cacheControl !== "undefined") {
            cacheObj = extra.handler.cacheControl;
        }
        var cacheControl = this._generateCacheHeader(cacheObj);
        if(cacheControl) {
            header["Cache-Control"] = cacheControl;
        }

        // Handle custom headers
        if(this.headers) {
           for(var key in this.headers) {
               header[key] = this.headers[key];
           }
        }
        if(extra && extra.handler && extra.handler.headers) {
            var endpointHeaders = extra.handler.headers;
            if(typeof endpointHeaders === "function") {
                endpointHeaders = endpointHeaders(data);
            }
            for(var key in endpointHeaders) {
                if(endpointHeaders === null) {
                    delete header[key];
                } else {
                    header[key] = endpointHeaders[key];
                }
            }
        }
        return header;
    },

    _generateCacheHeader: function(cacheObj) {
        var cacheHeaderParts = [];

        // Convert to an object if it is not
        if(cacheObj === false || cacheObj === 0 || this.debug) {
            // Cacheing disabled
            cacheObj = {noCache: true, noStore: true, mustRevalidate: true};
        } else if(typeof cacheObj === "number") {
            // Number which is treated as the maxAge
            cacheObj = {public: true, maxAge: cacheObj};
        } else if(typeof cacheObj === "string") {
            // Already a header string
            return cacheObj;
        } else if(typeof cacheObj !== "object" || cacheObj === null) {
            // Default
            cacheObj = {public: true, maxAge: 30};
        }
        
        // Where can it be cached
        if(cacheObj.public) {
            cacheHeaderParts.push('public');
        } else if(cacheObj.private) {
            cacheHeaderParts.push('private');
        } else if(cacheObj.noCache) {
            cacheHeaderParts.push('no-cache');
        } else if(cacheObj.onlyIfCached) {
            cacheHeaderParts.push('only-if-cached');
        }

        // Add all the other flags/paramiters
        if(cacheObj.noStore) {
            cacheHeaderParts.push('no-store');
        }
        if(cacheObj.mustRevalidate) {
            cacheHeaderParts.push('must-revalidate');
        }
        if(cacheObj.immutable) {
            cacheHeaderParts.push('immutable');
        }
        if(cacheObj.noTransform) {
            cacheHeaderParts.push('no-transform');
        }
        if(cacheObj.maxAge) {
            cacheHeaderParts.push('max-age='+cacheObj.maxAge);
        }
        if(cacheObj.sharedMaxAge) {
            cacheHeaderParts.push('s-max-age='+cacheObj.sharedMaxAge);
        }
        if(cacheObj.maxStale) {
            cacheHeaderParts.push('max-stale='+cacheObj.maxStale);
        }
        if(cacheObj.minFresh) {
            cacheHeaderParts.push('min-fresh='+cacheObj.minFresh);
        }
        if(cacheObj.staleWhileRevalidate) {
            cacheHeaderParts.push('stale-while-revalidate='+cacheObj.staleWhileRevalidate);
        }
        if(cacheObj.staleIfError) {
            cacheHeaderParts.push('stale-if-error='+cacheObj.staleIfError);
        }
        return cacheHeaderParts.join(', ');
    },

    _checkAuthDetails: function(handler_auth, authDetails, cb) {
        if(handler_auth) {
            return handler_auth(authDetails.user, authDetails.pass, cb);
        } else if(handler_auth === false) {
            // Authentication specifically disabled for this package
        } else if(this.checkAuth) {
            return this.checkAuth(authDetails.user, authDetails.pass, cb);
        }
        return cb(true);  // Default allow
    },

    /**
     * Writes the provided data into the response and closes the connection
     * @param res: Response object to write the data to
     * @param data: The data to reply with. Optionally data.code will be used as the response code and
     *                 data.error or data.msg will be used as the status message.
     * @returns: 
     */
    _writeResponse: function(res, data, extra) {
        if(!data) data = '' + data;
        if(!extra) extra = {};
        var header = this._getHeaders(data, extra);
        var statusMsg = extra.status || data.error || data.msg || 'success';
        var code = extra.code || data.code || 200;
        if(this.debug && extra.e && data.e === undefined) {
            data.e = extra.e.stack || (''+extra.e);
        }
        var data_string;
        if(this.prettyJson) {
            data_string = JSON.stringify(data, null, 2);
        } else {
            data_string = JSON.stringify(data);
        }

        this.logger.emit(this.info, 'Making ' + code + ' response', data);

        res.writeHead(code, statusMsg, header);
        return res.end(data_string);
    },

    _getRequestData: function(req) {
        if(this.fullRequest) {
            req.ip = req.connection.remoteAddress;
            return req;
        } else {
            return {
                method: req.method,
                args: req.args,
                body: req.body,
                ip: req.connection.remoteAddress
            }
        }
    },

    /**
     * Creates a handler for processing an incoming data packet
     * @param method: A string indicating the type of call we will handle, eg. 'GET' or 'POST'
     * @param getData: A function that given a request will return the provided data from the client
     * @returns: Returns a function that takes a request and response object and replies to the client
     */
    _process: function() {
        var self = this;
        return function(req, res) {
            try {
                self.middleware.run(req, res, function() {
                    if(req.method == 'GET') {
                        req.u = url.parse(req.url, true);
                        req.body = req.u.query;
                    } else {
                        req.u = url.parse(req.url, false);
                    }

                    var responseData = self._checkParamsSupplied(req);

                    if(!responseData.ok) {
                        return self._writeResponse(res, responseData, {handler: handler});
                    }

                    var handler = responseData.handler;
                    req.args = responseData.args;
                    var authDetails = Auth.decodeAuthDetails(req.headers.authorization);

                    self._checkAuthDetails(handler.auth, authDetails, function(auth) {
                        if(!auth) {
                            return self._writeResponse(res, {code: 401, ok: false, error: "Auth failed"}, {handler: handler});
                        }

                        // If we are still ok up to here then we can let the handler respond
                        setTimeout(function() {
                            try{
                                if(handler.length >= 2) {
                                    // Handler takes a callback so run asynchronously
                                    handler(self._getRequestData(req), function(err, result, extra) {
                                        extra = extra || {};
                                        extra[handler] = handler;
                                        if(err && extra.e === undefined) {
                                            extra.e = err;
                                        }
                                        self._writeResponse(res, result, extra);
                                    });
                                } else {
                                    // Handler doesn't take a callback so run synchronously
                                    var result = handler(self._getRequestData(req));
                                    self._writeResponse(res, result, {handler: handler});
                                }
                            } catch(e) {
                                self.logger.emit(self.warn, 'Uncaught exception in handler', {e: e});
                                self._writeResponse(res, {
                                    ok: false,
                                    code: 500,
                                    error: ''+e
                                }, {e: e});
                            }
                        }, 0);
                    });
                });
            } catch(e) {
                // Something has gone wrong! Reply to the user with a generic error
                self.logger.emit(self.error, 'Uncaught exception', {e: e, stack: e.stack});
                self._writeResponse(res, {
                    ok: false,
                    code: 500,
                    error: ''+e
                }, {e: e});
            }
        };
    },

    /**
     * Sets the global authentication function
     * @param f: The function to use to authenticate requests globally
     */
    auth: function(f) {
        this.checkAuth = f;
    },

    /**
     * Set the server to authenticate based on a dict of valid username->password mappings
     * @param credentials: A dict of username->password mappings, the password value can be a string or a list of strings
     */
    authByJson: function(credentials) {
        var auth_function = Auth.authByJsonFunction(credentials);
        // Convert to "async" function and set the server to auth using it
        this.auth(function(u, p, cb) {
            var r = auth_function(u, p);
            cb(r);
        });
    }
};

// Deprecated function names to be replaced in version 1.0.0
ApiQuick.prototype.getBasicHeader = ApiQuick.prototype._getHeaders;
ApiQuick.prototype.handleRateLimitHelper = function(ip) {
    return this.rateLimit.handleRateLimit(ip);
};

module.exports = new ApiQuick();
